File: dis-ingest/src/main/java/io/pivotal/dis/ingest/domain/Digest.java
package io.pivotal.dis.ingest.domain;


import java.util.List;
import java.util.Optional;
import java.util.function.Function;

public class Digest {
    private List<DisruptedLine> disruptions;

    public Digest(List<DisruptedLine> disruptions) {
        this.disruptions = disruptions;
    }

    public List<DisruptedLine> getDisruptions() {
        return disruptions;
    }

    public Optional<String> getStartTimeFromDisruptedLine(String lineName) {
        return getFieldFromDisruptedLine(lineName, DisruptedLine::getStartTime);
    }

    public Optional<String> getEndTimeFromDisruptedLine(String lineName) {
        return getFieldFromDisruptedLine(lineName, DisruptedLine::getEndTime);
    }

    public Optional<String> getEarliestEndTimeFromDisruptedLine(String lineName) {
        return getFieldFromDisruptedLine(lineName, DisruptedLine::getEarliestEndTime);
    }

    public Optional<String> getLatestEndTimeFromDisruptedLine(String lineName) {
        return getFieldFromDisruptedLine(lineName, DisruptedLine::getLatestEndTime);
    }

    private Optional<String> getFieldFromDisruptedLine(String lineName,
                                                       Function<DisruptedLine, String> fieldExtractor) {

        if (isLineDisrupted(lineName)) {
            DisruptedLine line = getLine(lineName).get();
            return Optional.of(fieldExtractor.apply(line));
        } else {
            return Optional.empty();
        }
    }

    public boolean isLineDisrupted(String lineName) {
        return disruptions.stream()
                .anyMatch(disruptedLine -> disruptedLine.getLine().equals(lineName));
    }

    public Optional<DisruptedLine> getLine(String lineName) {
        return getDisruptions().stream()
                .filter(disruptedLine -> disruptedLine.getLine().equals(lineName))
                .findFirst();
    }
}

File: dis-ingest/src/main/kotlin/io/pivotal/dis/ingest/domain/Digest.kt
package io.pivotal.dis.ingest.domain


import java.time.LocalDateTime
import java.util.*
import java.util.function.Function

class Digest(val disruptions: List<DisruptedLine>, val lastUpdated: Long) {

    private fun getLongFieldFromDisruptedLine(lineName: String,
                                              fieldExtractor: Function<DisruptedLine, Long>): Optional<Long> {

        if (isLineDisrupted(lineName)) {
            val line = getLine(lineName).get()
            return Optional.of(fieldExtractor.apply(line))
        } else {
            return Optional.empty<Long>()
        }
    }

    private fun isLineDisrupted(lineName: String): Boolean {
        return disruptions.any { disruptedLine -> disruptedLine.line == lineName }
    }

    private fun getLine(lineName: String): Optional<DisruptedLine> {
        return Optional.ofNullable(disruptions.filter { disruptedLine -> disruptedLine.line == lineName }.firstOrNull())
    }

    fun getStartTimestampFromDisruptedLine(lineName: String): Optional<Long> {
        return getLongFieldFromDisruptedLine(lineName, Function<DisruptedLine, Long> { it.startTimestamp })
    }

    fun getEndTimestampFromDisruptedLine(lineName: String): Optional<Long> {
        return getLongFieldFromDisruptedLine(lineName, Function<DisruptedLine, Long> { it.endTimestamp })
    }

    fun getEarliestEndTimestampFromDisruptedLine(lineName: String): Optional<Long> {
        return getLongFieldFromDisruptedLine(lineName, Function<DisruptedLine, Long> { it.earliestEndTimestamp })
    }

    fun getLatestEndTimestampFromDisruptedLine(lineName: String): Optional<Long> {
        return getLongFieldFromDisruptedLine(lineName, Function<DisruptedLine, Long> { it.latestEndTimestamp })
    }
}
